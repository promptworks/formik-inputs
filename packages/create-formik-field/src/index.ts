import * as React from "react";
import { useFormikContext, useField, FieldProps, FieldValidator } from "formik";

type AnyComponent = React.ComponentType<any>;

type BaseInputProps<T extends AnyComponent> = Omit<
  React.ComponentProps<T>,
  "value" | "onChange"
>;

interface FieldConfig<T extends AnyComponent> {
  name: string;
  validate?: FieldValidator;
  innerRef?: (instance: T | null) => void;
  validStyle?: any;
  validClassName?: string;
  validProps?: BaseInputProps<T>;
  invalidStyle?: any;
  invalidClassName?: string;
  invalidProps?: BaseInputProps<T>;
}

export type PropGetter<T extends AnyComponent> = (
  config: FieldProps
) => Partial<React.ComponentProps<T>>;

export type CustomFieldProps<T extends AnyComponent> = BaseInputProps<T> &
  FieldConfig<T>;

export interface CreateField<T extends AnyComponent> {
  component: T;
  displayName: string;
  getProps?: PropGetter<T>;
}

/**
 * Are we currently running in React Native?
 */
const isNative = navigator ? navigator.product === "ReactNative" : false;

/**
 * These are the default set of props that an input will receive.
 */
const getDefaultProps: PropGetter<any> = ({ form, field }) => ({
  onChange: (value: any) => form.setFieldValue(field.name, value),
  onBlur: () => form.setFieldTouched(field.name, true)
});

/**
 * Merge style props. On React native, we'll just return an array.
 */
const mergeStyles = (...styles: any[]) => {
  if (!styles.length) {
    return undefined;
  }

  if (isNative) {
    return styles.filter(Boolean);
  }

  return styles.reduce((a, b) => (b ? Object.assign(a, b) : a), {});
};

/**
 * Merge class names. On React Native, we'll always return undefined.
 */
const mergeClassNames = (...classNames: any[]) => {
  if (isNative) {
    return undefined;
  }

  return classNames.filter(Boolean).join(" ") || undefined;
};

/**
 * A higher-order component that will return a Formik-compatible form field.
 */
export const createField = <T extends AnyComponent>({
  displayName,
  getProps = getDefaultProps,
  component: WrappedComponent
}: CreateField<T>) => {
  const FormikField: React.FC<CustomFieldProps<T>> = ({
    name,
    validate,
    innerRef,
    style,
    className,
    validClassName = isNative ? undefined : "is-valid",
    validStyle,
    validProps,
    invalidClassName = isNative ? undefined : "is-invalid",
    invalidStyle,
    invalidProps,
    ...props
  }) => {
    const formik = useFormikContext<any>();
    const [field, meta] = useField(name);

    // FIXME: This will happen automatically when this PR is released:
    // https://github.com/jaredpalmer/formik/pull/1699
    React.useEffect(() => {
      formik.registerField(name, { validate });
      return () => formik.unregisterField(name);
    }, [formik.registerField, formik.unregisterField, name, validate]);

    const isValid = meta.touched && !meta.error;
    const isInvalid = meta.touched && meta.error;

    const inputProps = React.useMemo(
      () => getProps({ form: formik, field, meta }),
      [formik, field, meta]
    );

    return React.createElement(WrappedComponent, {
      ref: innerRef,
      value: field.value,
      style: mergeStyles(
        style,
        isValid && validStyle,
        isInvalid && invalidStyle
      ),
      className: mergeClassNames(
        className,
        isValid && validClassName,
        isInvalid && invalidClassName
      ),
      ...inputProps,
      ...props,
      ...(isValid ? validProps : {}),
      ...(isInvalid ? invalidProps : {})
    });
  };

  // Set a display name for the input
  FormikField.displayName = displayName;

  return FormikField;
};
