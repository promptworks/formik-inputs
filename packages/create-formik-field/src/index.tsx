import React from "react";
import { useFormikContext, useField, FieldProps, FieldValidator } from "formik";

type AnyComponent = React.ComponentType<any>;

type BaseInputProps<T extends AnyComponent> = Omit<
  React.ComponentProps<T>,
  "value" | "onChange"
>;

interface FieldConfig<T extends AnyComponent> {
  name: string;
  validate?: FieldValidator;
  innerRef?: (instance: T | null) => void;
}

export type PropGetter<T extends AnyComponent> = (
  config: FieldProps
) => Partial<React.ComponentProps<T>>;

export type CustomFieldProps<T extends AnyComponent> = BaseInputProps<T> &
  FieldConfig<T>;

export interface CreateField<T extends AnyComponent> {
  component: T;
  displayName: string;
  getProps: PropGetter<T>;
}

export const createField = <T extends AnyComponent>({
  getProps,
  displayName,
  component: WrappedComponent
}: CreateField<T>) => {
  const FormikField: React.FC<CustomFieldProps<T>> = ({
    name,
    validate,
    innerRef,
    ...props
  }) => {
    const formik = useFormikContext<any>();
    const [field, meta] = useField(name);

    // FIXME: This will happen automatically when this PR is released:
    // https://github.com/jaredpalmer/formik/pull/1699
    React.useEffect(() => {
      formik.registerField(name, { validate });
      return () => formik.unregisterField(name);
    }, [formik.registerField, formik.unregisterField, name, validate]);

    const inputProps = React.useMemo(
      () => getProps({ form: formik, field, meta }),
      [formik, field, meta]
    );

    return React.createElement(WrappedComponent, {
      ref: innerRef,
      value: field.value,
      ...inputProps,
      ...props
    });
  };

  // Set a display name for the input
  FormikField.displayName = displayName;

  return FormikField;
};
