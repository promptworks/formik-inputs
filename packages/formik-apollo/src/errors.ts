import { GraphQLError } from "graphql";
import { ApolloError, isApolloError } from "apollo-client";
import { Problems, ValidationError } from "./types";

const isString = (value: any) => typeof value === "string";
const isError = (value: any) => value instanceof Error;
const isObject = (value: any) => value && typeof value === "object";

/**
 * Indicates that an error is a validation error.
 */
export const VALIDATION_ERROR = "VALIDATION_ERROR";

/**
 * Create a validation error. Useful for testing.
 */
export const createValidationError = (message: string, problems: Problems) => {
  const gqlError = new GraphQLError(
    message,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    { code: VALIDATION_ERROR, problems }
  );

  return new ApolloError({
    errorMessage: message,
    graphQLErrors: [gqlError]
  });
};

/**
 * Get the code for an error.
 */
export const getErrorCode = (error: Error): string | undefined =>
  error instanceof GraphQLError &&
  isObject(error.extensions) &&
  error.extensions!.code;

/**
 * Is the error a validation error?
 */
export const isValidationError = (error: Error): error is ValidationError =>
  error instanceof GraphQLError &&
  isObject(error.extensions) &&
  isObject(error.extensions!.problems) &&
  error.extensions!.code === VALIDATION_ERROR;

/**
 * Extract all validation errors from an error.
 */
export const getValidationErrors = (
  error: ApolloError | GraphQLError | Error
): Record<string, string> => {
  if (!isError(error)) {
    return {};
  }

  if (isApolloError(error)) {
    return error.graphQLErrors.reduce(
      (acc, e) => Object.assign(acc, getValidationErrors(e)),
      {}
    );
  }

  if (!isValidationError(error)) {
    return {};
  }

  return Object.entries(error.extensions.problems).reduce(
    (acc, [key, value]) => {
      const message = Array.isArray(value) ? value[0] : value;
      const updates = isString(message) ? { [key]: message } : {};
      return Object.assign(acc, updates);
    },
    {}
  );
};
