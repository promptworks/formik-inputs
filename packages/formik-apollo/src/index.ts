export * from "./errors";
export * from "./useSubmit";
export * from "./types";
