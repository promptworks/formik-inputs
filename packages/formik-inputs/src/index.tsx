import * as Baseline from "react-baseline-inputs";
import { CustomFieldProps, createCustomField } from "./createCustomField";

export type InputProps = CustomFieldProps<typeof Baseline.Input>;
export const Input = createCustomField(Baseline.Input, "FormikInput");

export type CheckboxProps = CustomFieldProps<typeof Baseline.Checkbox>;
export const Checkbox = createCustomField(Baseline.Checkbox, "FormikCheckbox");

export type SelectProps = CustomFieldProps<typeof Baseline.Select>;
export const Select = createCustomField(Baseline.Select, "FormikSelect");

export type IntegerInputProps = CustomFieldProps<typeof Baseline.IntegerInput>;
export const IntegerInput = createCustomField(
  Baseline.IntegerInput,
  "FormikIntegerInput"
);

export type FloatInputProps = CustomFieldProps<typeof Baseline.FloatInput>;
export const FloatInput = createCustomField(
  Baseline.FloatInput,
  "FormikFloatInput"
);

export type FileInputProps = CustomFieldProps<typeof Baseline.FileInput>;
export const FileInput = createCustomField(
  Baseline.FileInput,
  "FormikFileInput"
);

export type DateInputProps = CustomFieldProps<typeof Baseline.DateInput>;
export const DateInput = createCustomField(
  Baseline.DateInput,
  "FormikDateInput"
);

export type DateTimeInputProps = CustomFieldProps<
  typeof Baseline.DateTimeInput
>;
export const DateTimeInput = createCustomField(
  Baseline.DateTimeInput,
  "FormikDateTimeInput"
);
