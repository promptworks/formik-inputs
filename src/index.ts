import { FieldProps } from "formik";
import * as Baseline from "react-native-baseline-inputs";
import { createField } from "./createField";
import { CustomFieldProps } from "./types";

const getInputProps = ({ form, field }: FieldProps) => ({
  onChange: (value: any) => form.setFieldValue(field.name, value),
  onBlur: () => form.setFieldTouched(field.name, true)
});

const getFallbackProps = ({ form, field }: FieldProps) => ({
  onChange: (value: any) => {
    form.setFieldValue(field.name, value);
    form.setFieldTouched(field.name, true);
  }
});

/**
 * Input
 */
export type InputProps = CustomFieldProps<typeof Baseline.Input>;
export const Input = createField<typeof Baseline.Input>({
  displayName: "FormikInput",
  component: Baseline.Input,
  getProps: getInputProps
});

/**
 * Switch
 */
export type SwitchProps = CustomFieldProps<typeof Baseline.Switch>;
export const Switch = createField<typeof Baseline.Switch>({
  displayName: "FormikSwitch",
  component: Baseline.Switch,
  getProps: getFallbackProps
});

/**
 * Picker
 */
export type PickerProps = CustomFieldProps<typeof Baseline.Picker>;
export const Picker = createField<typeof Baseline.Picker>({
  displayName: "FormikPicker",
  component: Baseline.Picker,
  getProps: getFallbackProps
});

/**
 * Numeric input
 */
export type NumericInputProps = CustomFieldProps<typeof Baseline.NumericInput>;
export const NumericInput = createField<typeof Baseline.NumericInput>({
  displayName: "FormikNumericInput",
  component: Baseline.NumericInput,
  getProps: getInputProps
});

/**
 * Integer input
 */
export type IntegerInputProps = CustomFieldProps<typeof Baseline.IntegerInput>;
export const IntegerInput = createField<typeof Baseline.IntegerInput>({
  displayName: "FormikIntegerInput",
  component: Baseline.IntegerInput,
  getProps: getInputProps
});

/**
 * Float input
 */
export type FloatInputProps = CustomFieldProps<typeof Baseline.FloatInput>;
export const FloatInput = createField<typeof Baseline.FloatInput>({
  displayName: "FormikFloatInput",
  component: Baseline.FloatInput,
  getProps: getInputProps
});

/**
 * Date picker
 */
export type DatePickerProps = CustomFieldProps<typeof Baseline.DatePicker>;
export const DatePicker = createField<typeof Baseline.DatePicker>({
  displayName: "FormikDatePicker",
  component: Baseline.DatePicker,
  getProps: getFallbackProps
});

/**
 * Time picker
 */
export type TimePickerProps = CustomFieldProps<typeof Baseline.TimePicker>;
export const TimePicker = createField<typeof Baseline.TimePicker>({
  displayName: "FormikTimePicker",
  component: Baseline.TimePicker,
  getProps: getFallbackProps
});

/**
 * DateTime picker
 */
export type DateTimePickerProps = CustomFieldProps<
  typeof Baseline.DateTimePicker
>;
export const DateTimePicker = createField<typeof Baseline.DateTimePicker>({
  displayName: "FormikDateTimePicker",
  component: Baseline.DateTimePicker,
  getProps: getFallbackProps
});
