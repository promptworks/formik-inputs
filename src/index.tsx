import * as React from "react";
import * as Inputs from "react-baseline-inputs";
import { FieldValidator, useField } from "formik";

type InputCreator<P> = <T>(
  theme: Inputs.Theme<T, P>
) => React.ComponentType<T & P>;

export interface FormikThemeProps {
  touched?: boolean;
}

export type FormikInput<P> = Omit<P, "name" | "value" | "onChangeValue"> & {
  name: string;
  validate?: FieldValidator;
};

function getName(component: any): string {
  return component.displayName || component.name || "Field";
}

function convert<P>(createBaselineField: InputCreator<P>) {
  return function createField<T extends FormikThemeProps>(
    theme: Inputs.Theme<T, P>
  ) {
    const BaselineField = createBaselineField({
      ...theme,
      props: [...(theme.props || []), "touched"]
    });

    function Field(props: FormikInput<P>) {
      const { name, validate, ...otherProps } = props;
      const [field, meta, helpers] = useField({ name, validate });

      const fieldProps: any = {
        name: name,
        value: field.value,
        touched: meta.touched,
        error: meta.error,
        onBlur: field.onBlur,
        onChangeValue: helpers.setValue,
        ...otherProps
      };

      return <BaselineField {...fieldProps} />;
    }

    Field.displayName = `Formik${getName(BaselineField)}`;

    return Field;
  };
}

export const createCheckbox = convert(Inputs.createCheckbox);
export const createFileInput = convert(Inputs.createFileInput);
export const createFileListInput = convert(Inputs.createFileListInput);
export const createGroup = convert(Inputs.createGroup);
export const createInput = convert(Inputs.createInput);
export const createItem = Inputs.createItem;
export const createSelect = convert(Inputs.createSelect);
export const createTextArea = convert(Inputs.createTextArea);
export const createToggleButton = convert(Inputs.createToggleButton);

export const Checkbox = createCheckbox({});
export const FileInput = createFileInput({});
export const FileListInput = createFileListInput({});
export const Group = createGroup({});
export const Input = createInput({});
export const Item = Inputs.Item;
export const Select = createSelect({});
export const TextArea = createTextArea({});
export const ToggleButton = createToggleButton({});
