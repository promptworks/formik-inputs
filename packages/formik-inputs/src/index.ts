import * as Baseline from "react-baseline-inputs";
import { createField, CustomFieldProps, PropGetter } from "create-formik-field";

export type TextInputProps = CustomFieldProps<typeof Baseline.TextInput>;
export type TextAreaProps = CustomFieldProps<typeof Baseline.TextArea>;
export type CheckboxProps = CustomFieldProps<typeof Baseline.Checkbox>;
export type SelectProps = CustomFieldProps<typeof Baseline.Select>;
export type IntegerInputProps = CustomFieldProps<typeof Baseline.IntegerInput>;
export type FloatInputProps = CustomFieldProps<typeof Baseline.FloatInput>;
export type FileInputProps = CustomFieldProps<typeof Baseline.FileInput>;
export type DateInputProps = CustomFieldProps<typeof Baseline.DateInput>;
export type DateTimeInputProps = CustomFieldProps<
  typeof Baseline.DateTimeInput
>;

const getInputProps: PropGetter<any> = ({ form, field }) => ({
  name: field.name,
  onChange: (value: any) => form.setFieldValue(field.name, value),
  onBlur: () => form.setFieldTouched(field.name, true)
});

/**
 * An HTML `<input />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It defaults to `type="text"`.
 *   * It emits a `null` value to the `onChange` handler when the input is empty.
 */
export const TextInput = createField<typeof Baseline.TextInput>({
  component: Baseline.TextInput,
  displayName: "FormikTextInput",
  getProps: getInputProps
});

/**
 * An HTML `<textarea />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits a `null` value to the `onChange` handler when the input is empty.
 */
export const TextArea = createField<typeof Baseline.TextArea>({
  component: Baseline.TextArea,
  displayName: "FormikTextArea",
  getProps: getInputProps
});

/**
 * An HTML `<input type='checkbox' />`, but with the following benefits:
 *
 *   * It accepts boolean values.
 *   * It accepts `null` as a value.
 *   * It always emits a boolean value when changed.
 */
export const Checkbox = createField({
  component: Baseline.Checkbox,
  displayName: "FormikCheckbox",
  getProps: getInputProps
});

/**
 * An HTML `<select />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It accepts a `placeholder`.
 *   * It allows you to pass in an array of options.
 */
export const Select = createField({
  component: Baseline.Select,
  displayName: "FormikSelect",
  getProps: getInputProps
});

/**
 * An HTML `<input type="number" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It always emits an integer when changed.
 */
export const IntegerInput = createField({
  component: Baseline.IntegerInput,
  displayName: "FormikIntegerInput",
  getProps: getInputProps
});

/**
 * An HTML `<input type="number" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It always emits a number when changed.
 */
export const FloatInput = createField({
  component: Baseline.FloatInput,
  displayName: "FormikFloatInput",
  getProps: getInputProps
});

/**
 * An HTML `<input type="file" />`, but with the following benefits:
 *
 *   * It emits a `File | null` when changed.
 *   * When `multiple`, it will emit a `FileList` when changed.
 *   * It ignores any `value` prop that you give it.
 */
export const FileInput = createField({
  component: Baseline.FileInput,
  displayName: "FormikFileInput",
  getProps: getInputProps
});

/**
 * An HTML `<input type="date" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits an ISO-formatted date string when changed.
 */
export const DateInput = createField({
  component: Baseline.DateInput,
  displayName: "FormikDateInput",
  getProps: getInputProps
});

/**
 * An HTML `<input type="datetime-local" />`, but with the following benefits:
 *
 *   * It accepts `null` as a value.
 *   * It emits an ISO-formatted datetime string when changed.
 */
export const DateTimeInput = createField({
  component: Baseline.DateTimeInput,
  displayName: "FormikDateTimeInput",
  getProps: getInputProps
});
