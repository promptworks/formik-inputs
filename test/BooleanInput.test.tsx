import React from "react";
import { shallow } from "enzyme";
import { BooleanInput } from "../src";

it("renders", () => {
  const onChange = jest.fn();
  const input = shallow(<BooleanInput value={true} onChange={onChange} />);
  expect(input).toMatchSnapshot();
});

it("emits a boolean value on change", () => {
  const onChange = jest.fn();
  const input = shallow(<BooleanInput value={true} onChange={onChange} />);
  input.find("Switch").simulate("valueChange", false);
  expect(onChange).toHaveBeenCalledWith(false);
});
