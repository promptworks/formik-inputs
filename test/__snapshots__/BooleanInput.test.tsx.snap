// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Switch
  onValueChange={[MockFunction]}
  value={true}
/>
`;
